frame/treasury/src/lib.rs

```diff
@@ -124,10 +295,188 @@ decl_module! {
+			Self::deposit_event(RawEvent::TipRetracted(hash));
+		}
+
+		/// Give a tip for something new; no finder's fee will be taken.
+		///
+		/// The dispatch origin for this call must be _Signed_ and the signing account must be a
+		/// member of the `Tippers` set.
+		///
+		/// - `reason`: The reason for, or the thing that deserves, the tip; generally this will be
+		///   a UTF-8-encoded URL.
+		/// - `who`: The account which should be credited for the tip.
+		/// - `tip_value`: The amount of tip that the sender would like to give. The median tip
+		///   value of active tippers will be given to the `who`.
+		///
+		/// Emits `NewTip` if successful.
+		///
+		/// # <weight>
+		/// - `O(R + T)` where `R` length of `reason`, `T` is the number of tippers. `T` is
+		///   naturally capped as a membership set, `R` is limited through transaction-size.
+		/// - Two storage insertions (codecs `O(R)`, `O(T)`), one read `O(1)`.
+		/// - One event.
+		/// # </weight>
+		#[weight = SimpleDispatchInfo::FixedNormal(150_000)]
+		fn tip_new(origin, reason: Vec<u8>, who: T::AccountId, tip_value: BalanceOf<T>) {
+			let tipper = ensure_signed(origin)?;
+			ensure!(T::Tippers::contains(&tipper), BadOrigin);
+			let reason_hash = T::Hashing::hash(&reason[..]);
+			ensure!(!Reasons::<T>::exists(&reason_hash), Error::<T>::AlreadyKnown);
+			let hash = T::Hashing::hash_of(&(&reason_hash, &who));
+
+			Reasons::<T>::insert(&reason_hash, &reason);
+			Self::deposit_event(RawEvent::NewTip(hash.clone()));
+			let tips = vec![(tipper, tip_value)];
+			let tip = OpenTip { reason: reason_hash, who, finder: None, closes: None, tips };
+			Tips::<T>::insert(&hash, tip);
+		}
+
+		/// Declare a tip value for an already-open tip.
+		///
+		/// The dispatch origin for this call must be _Signed_ and the signing account must be a
+		/// member of the `Tippers` set.
+		///
+		/// - `hash`: The identity of the open tip for which a tip value is declared. This is formed
+		///   as the hash of the tuple of the hash of the original tip `reason` and the beneficiary
+		///   account ID.
+		/// - `tip_value`: The amount of tip that the sender would like to give. The median tip
+		///   value of active tippers will be given to the `who`.
+		///
+		/// Emits `TipClosing` if the threshold of tippers has been reached and the countdown period
+		/// has started.
+		///
+		/// # <weight>
+		/// - `O(T)`
+		/// - One storage mutation (codec `O(T)`), one storage read `O(1)`.
+		/// - Up to one event.
+		/// # </weight>
+		#[weight = SimpleDispatchInfo::FixedNormal(50_000)]
+		fn tip(origin, hash: T::Hash, tip_value: BalanceOf<T>) {
+			let tipper = ensure_signed(origin)?;
+			ensure!(T::Tippers::contains(&tipper), BadOrigin);
+
+			let mut tip = Tips::<T>::get(hash).ok_or(Error::<T>::UnknownTip)?;
+			if Self::insert_tip_and_check_closing(&mut tip, tipper, tip_value) {
+				Self::deposit_event(RawEvent::TipClosing(hash.clone()));
+			}
+			Tips::<T>::insert(&hash, tip);
```

Comment:
Sorry for comment with this historical PR, but I want to ask were there any special considerations without emitting a `Tip` event here? Just to avoid too many events?

Actually it's a little troublesome without this event to handle tip off-chain.

Reply:
I think we can emit some events if needed, it shouldn't be too may events AFAICT. You can open an issue or PR

Reply:
> I think we can emit some events if needed, it shouldn't be too may events AFAICT. You can open an issue or PR

I see. Thanks.